blog/_src/posts/2018-08-16-if-you-give-a-mouse-a-macro.md

```diff
@@ -0,0 +1,445 @@
+includes the necessary information as part of its syntax. Then, the surrounding
+context can analyze the expansion of each term, looking for that form.
+
+Concretely, `define` will expand to `define/intermediate`, which will in turn
+expand to what `define` originally expanded to:
+
+```racket
+(define-typed-syntax define
+  [(_ x:id e) ≫
+   [⊢ e ≫ e- ⇒ τ]
+   #:with x- (generate-temporary #'x)
+   #:with x+ (syntax-local-identifier-as-binding #'x)
+   -----------------------------------------------------
+   [⊢ (define/intermediate x+ x- τ e-) ⇒ Void]]
+  [(_ (f [x (~datum :) τ] ...) e ...+) ≫
+   -----------------------------------
+   [≻ (define f (λ+ ([x : τ] ...) e ...))]])         
+
+(define-syntax (define/intermediate stx)
+  (syntax-parse stx
+    [(_ x:id x-:id τ e)
+     #'(begin-
+         (define-typed-variable-rename x ≫ x- : τ)
+         (define- x- e)
+         a-deep-dark-void)]))
+```
+
+The reason we create an `x+` using [`syntax-local-identifier-as-binding`][13] is
+[due to a bug in the expander][12]. The explanation is rather involved and
+frankly I only barely understand what's going on myself (if at all), so let's
+just leave it at that and move on.
+
+Then, for each form `e` in a sequence, we can call `local-expand` with `def-ctx`
+and then check the expansion, `e-`, for `define/intermediate`. In those cases,
+we can use `int-def-ctx-bind-type-rename` to add it to the context. The
+procedure `add-bindings-to-ctx` performs this check on an expanded form `e-`:
+
+```racket
+(define-for-syntax (add-bindings-to-ctx e- def-ctx)
+  (syntax-parse e-
+        #:literals (erased define/intermediate)
+        [(erased (define/intermediate x:id x-:id τ e-))
+         (int-def-ctx-bind-type-rename #'x #'x- #'τ def-ctx)]
+        [_ (void)]))
+        
+```
+
+One extra detail that snuck in to the implementation is the presence of the
+`erased` tag. `erased` is a macro inserted by Turnstile that serves to mark
+language forms that have already been expanded, helping control what expands
+when. Since we used `define-typed-syntax` for our `define`, Turnstile wraps the
+output with `erased`.
+
+We now have the key ingredients to define a procedure, `walk/bind` that will
+serve as the primary vehicle to perform type checking of a sequence of forms,
+threading binding information through using a definition context. Processing
+sequences of defintions and expressions will iterate through them one at a time,
+and for each form `e`:
+
+- `local-expand` `e` using our internal definition context, resulting in an `e-`.
+- Retrieve the type of `e` from the metadata of `e-` using Turnstile's `typeof` helper.
+- Check if `e` defined a binding, in which case add it to the context.
+
+Aggregating the expanded syntax and type of each form as we go along, we get
+
+```racket
+(define-for-syntax (walk/bind e...)
+  (define def-ctx (syntax-local-make-definition-context))
+  (define unique (gensym 'walk/bind))
+  (define-values (rev-e-... rev-τ...)
```

Comment:
In this case I deliberately made an identifier named `e...` (and other similar ones) because they correspond to the appearances of `e ...` in the related macros. Is that too much abuse?